framework/src/executioners/SecantSolve.C

```diff
@@ -0,0 +1,271 @@
+  for (size_t i = 0; i < _secondary_transformed_pps.size(); i++)
+  {
+    // Advance one step
+    _secondary_transformed_pps_values[i][3] = _secondary_transformed_pps_values[i][1];
+    _secondary_transformed_pps_values[i][2] = _secondary_transformed_pps_values[i][0];
+
+    // Secondary postprocessors are saved after a solve, but transfers from the main app have not
+    // been received yet
+    _secondary_transformed_pps_values[i][1] =
+        getPostprocessorValueByName(_secondary_transformed_pps[i]);
+  }
+}
+
+void
+SecantSolve::savePreviousValuesAsMainApp()
+{
+  // Save previous variable values
+  NumericVector<Number> & solution = _nl.solution();
+  NumericVector<Number> & fxn_m1 = _nl.getVector("fxn_m1");
+  NumericVector<Number> & xn_m1 = _nl.getVector("xn_m1");
+  NumericVector<Number> & fxn_m2 = _nl.getVector("fxn_m2");
+  NumericVector<Number> & xn_m2 = _nl.getVector("xn_m2");
+
+  // Advance one step
+  xn_m2 = xn_m1;
+  fxn_m2 = fxn_m1;
+
+  // Before a solve (here), solution is a sequence term, after a solve, solution is the evaluated
+  // term
+  xn_m1 = solution;
+
+  // Set postprocessor previous values
+  for (size_t i = 0; i < _transformed_pps.size(); i++)
+  {
+    // Advance one step
+    _transformed_pps_values[i][3] = _transformed_pps_values[i][1];
+    _transformed_pps_values[i][2] = _transformed_pps_values[i][0];
+
+    // Primary postprocessors are saved before a solve
+    _transformed_pps_values[i][1] = getPostprocessorValueByName(_transformed_pps[i]);
+  }
+}
+
+bool
+SecantSolve::useFixedPointAlgorithmUpdate(bool as_main_app)
+{
+  // Need at least two evaluations to compute the Secant slope
+  if (as_main_app)
+    return _fixed_point_it > 1;
+  else
+    return _main_fixed_point_it > 1;
+}
+
+void
+SecantSolve::transformPostprocessorsAsMainApp()
+{
+  // Relax postprocessors for the main application
+  for (size_t i = 0; i < _transformed_pps.size(); i++)
+  {
+    // Get new postprocessor value
+    const Real current_value = getPostprocessorValueByName(_transformed_pps[i]);
+    const Real xn_m1 = _transformed_pps_values[i][1];
+    const Real fxn_m2 = _transformed_pps_values[i][2];
+    const Real xn_m2 = _transformed_pps_values[i][3];
+
+    // Save fxn_m1, received or computed before the solve
+    _transformed_pps_values[i][0] = current_value;
+
+    // Compute and set relaxed value
+    Real new_value = current_value;
+    if (!MooseUtils::absoluteFuzzyEqual(current_value - xn_m1 - fxn_m2 + xn_m2, 0))
+      new_value = xn_m1 - (current_value - xn_m1) * (xn_m1 - xn_m2) /
+                              (current_value - xn_m1 - fxn_m2 + xn_m2);
+
+    // Relax update if desired
+    new_value = _relax_factor * new_value + (1 - _relax_factor) * xn_m1;
+
+    _problem.setPostprocessorValueByName(_transformed_pps[i], new_value);
+  }
+}
+
+void
+SecantSolve::transformPostprocessorsAsSubApp()
+{
+  // Update the postprocessors
+  for (size_t i = 0; i < _secondary_transformed_pps.size(); i++)
+  {
```

Comment:
These two functions seem to share a lot of implementation.  Also all the allocateStorage... savePrevious... transform... functions really seem to have a lot of redundant feeling to them - like boiler plate that doesn't quite feel right design-wise.  Also the transform... functions really are the "main" functions of these objects/algorithms - so it feels like these should be named like computeSolutionUpdate or something more evoking of their importance.

I'm not sure what the right solution is, but this part of your API design doesn't feel quite right.  Maybe you could try to explain some of the constraints/issues you ran into that lead you to designing it this way?  The asSubApp and asMainApp diversion is a bit odd - if feels like the algorithm shouldn't care about that - maybe we pass in the primary vs secondary vectors/pps as arguments and the algorithm just blindly operates on what it is given instead of having the two separate functions.  That may not really make sense/work, I'm just spitballing here...  Mostly the main vs sub app bifurcation just seems rough and requires a lot of manual "doing the right thing" - instead of allowing the code to do things the right way for us.

Reply:
That's a good idea. I could probably consolidate the transform, maybe the save, functions. We'll loose some explicitness as to what is being updated/transformed but it'll be easier to maintain each method. 
Part of the reason why these functions were separated was how they came about, and that's no longer a constraint in the current shape of the code.

I like transformed more than updated. Like the action of the method is to transform the sequence.